File: backend/app/schemas/analytics.py
from datetime import date
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel


class PerformancePeriod(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    YEARLY = "yearly"


class PerformancePoint(BaseModel):
    period_end: date
    invested_value: Decimal
    market_value: Decimal | None
    """
    market_value is only populated for the most recent point (today), since
    this schema has no historical price table. Earlier points reflect
    cumulative invested capital from the transaction ledger only.
    """


class PerformanceHistoryResponse(BaseModel):
    period: PerformancePeriod
    points: list[PerformancePoint]


class AllocationItem(BaseModel):
    symbol: str
    company_name: str
    value: Decimal
    percent: Decimal


class AssetAllocationResponse(BaseModel):
    items: list[AllocationItem]
    total_value: Decimal


class SectorAllocationItem(BaseModel):
    sector: str
    value: Decimal
    percent: Decimal


class SectorAllocationResponse(BaseModel):
    items: list[SectorAllocationItem]
    total_value: Decimal


class TopMoverItem(BaseModel):
    symbol: str
    company_name: str
    current_price: Decimal
    unrealized_pnl: Decimal
    unrealized_pnl_percent: Decimal


class TopMoversResponse(BaseModel):
    gainers: list[TopMoverItem]
    losers: list[TopMoverItem]


class PortfolioMetrics(BaseModel):
    total_invested_value: Decimal
    total_market_value: Decimal
    total_unrealized_pnl: Decimal
    total_unrealized_pnl_percent: Decimal
    holdings_count: int
    sectors_count: int
    best_performer: TopMoverItem | None
    worst_performer: TopMoverItem | None

File: frontend/src/types/analytics.ts
export type PerformancePeriod = "daily" | "weekly" | "monthly" | "yearly";

export interface PerformancePoint {
  period_end: string;
  invested_value: string;
  market_value: string | null;
}

export interface PerformanceHistoryResponse {
  period: PerformancePeriod;
  points: PerformancePoint[];
}

export interface AllocationItem {
  symbol: string;
  company_name: string;
  value: string;
  percent: string;
}

export interface AssetAllocationResponse {
  items: AllocationItem[];
  total_value: string;
}

export interface SectorAllocationItem {
  sector: string;
  value: string;
  percent: string;
}

export interface SectorAllocationResponse {
  items: SectorAllocationItem[];
  total_value: string;
}

export interface TopMoverItem {
  symbol: string;
  company_name: string;
  current_price: string;
  unrealized_pnl: string;
  unrealized_pnl_percent: string;
}

export interface TopMoversResponse {
  gainers: TopMoverItem[];
  losers: TopMoverItem[];
}

export interface PortfolioMetrics {
  total_invested_value: string;
  total_market_value: string;
  total_unrealized_pnl: string;
  total_unrealized_pnl_percent: string;
  holdings_count: number;
  sectors_count: number;
  best_performer: TopMoverItem | null;
  worst_performer: TopMoverItem | null;
}
